### PR TITLE
Request gap refactor

### PR DESCRIPTION
**Pull request by csciguy8**

### CODE REVIEW

**Comment by csciguy8** [comment]
All major restructuring is done and have posted (in the PR description) the latest Google P3DT performance numbers using Cesium for Unreal performance tests.

A snippet...
```
GoogleTiles.LocaleDeathValley (cold cache) - 25% load time improvement
GoogleTiles.LocaleDeathValley (warm cache) - 44% load time slow down
GoogleTiles.LocaleChrysler (cold cache) - 17% load time improvement
GoogleTiles.LocaleChrysler (warm cache) - 48% load time slow down
```

@kring , keeping you in the loop. Oddly enough, looks like we're still seeing roughly the same improvements as originally guessed back in November, when the original issue was written, 17%-25% improvements for Google P3DT scenes in Unreal. And we're doing that without increasing the max simultaneous loads (still 20).

Unfortunately, the warm cache numbers got worse. The reason is not obvious, but there may be something we can do to make this better. Will add a check box to the TO DOs.

**Comment by kring** [comment]
This is a total guess, but one thing to watch out for is that the warm cache case is especially suceptible to slowdowns caused by waiting to run tasks. For example, a simple `thenInMainThread` somewhere in the load processing can cause a delay of 16ms (at 60FPS) before that part of the task can run, which is quite a long time with a warm cache.

**Comment by csciguy8** [comment]
Finished performance testing and updated the missing numbers.

A snippet
```
SampleLocaleDenver (cold cache) - 16% load time slow down
SampleLocaleDenver (warm cache) -22% load time slow down
SampleLocaleMelbourne (cold cache) - about the same
SampleLocaleMelbourne (warm cache) - about the same
```

A little odd that Denver slowed down a bit. It is our smallest test, and already pretty fast. First guess is that this PR is indeed batching content requests better as shown for the google levels. But for cached (or very fast) content requests, it may be introducing delay (or more overhead) for processing. Like @kring said, this could be an errant `thenInMainThread`, or something similar that is fixable.

I think this is ready for review and to talk a bit more about performance numbers. 

Making this a non-draft PR...

**Comment by csciguy8** [comment]
I think I have a solid lead for the slowdown in the Google warm cache case... 

If I turn off the time budget in `Tileset::_processMainThreadLoadQueue`, the loading times in Google Chrysler (warm) between main and this PR become almost the same (within .1s). 

This function is throttling the final stage of tile loading based on a time budget, to keep a reasonable frame rate. 

In the context of this PR, we can think of it as the final 'stage' in our staged loading pipeline. And taking into account what this PR is doing, it seems likely that it's pushing the same amount of work through this stage, but with bigger batches of work, later in time, ultimately leading to longer load times.

Will think on this a bit more...

**Comment by csciguy8** [comment]
Found the slowdown in the Google warm cache case! (finally).

TLDR, the reference branch (based on main) I'm comparing to had a bug where it completed too early.  Fixing the bug and doing a quick retest showed me that this PR doesn't seem to introduce any weird slow downs in this case anymore. Will retest to verify.

I've fixed it in my branch [here](https://github.com/CesiumGS/cesium-native/commit/f34915b8884fe5d3441f1b220111009fd4ba0052), but it could optionally be its own PR as well, since it's a bug in main too.

More in depth, the problem is in `::computeLoadProgress` and is reporting 100% done before we are actually done. It uses this code to determine if there's any work in the any of our loading queues...
`int32_t queueSizeSum = static_cast<int32_t>(
      this->_workerThreadLoadQueue.size() + this->_mainThreadLoadQueue.size());`

Unfortunately, `_mainThreadLoadQueue` gets cleared every frame, so this is always 0, and any progress that is computed is ignoring any main thread loading that happened, or was ever queued.

This behavior coincides with what I saw in the original tests. Inexplicably fast performance compared to my PR, where I had inadvertently fixed this bug through refactoring. It also coincides with my observation of equal performance when turning off time budget. Doing this clears the main thread queue every frame, which essentially makes it always 0 when `::computeLoadProgress` is called.

**Comment by csciguy8** [comment]
Another update to the performance test numbers...

Good news: the performance wins from this PR went up! (now 25-32% improvements for the google tests)

Mediocre news: although the inexplicable "warm cache is really slow" problem is gone, warm cache results are still slightly slower in this PR (9-21% slower). Denver cold cache falls into this category too. I have a pretty good reason to believe that the problem is [here](https://github.com/CesiumGS/cesium-native/blob/a16c3d5fa6ef974b2faede193f7cca80cba1e741/Cesium3DTilesSelection/src/TileWorkManager.cpp#L476), which coincides with @kring 's recommendation to look out for main thread continuations. 

The fix would be to dispatch processing work immediately, rather than waiting to dispatch on the main thread. Not a trivial change, but not an unreasonable amount of work either.

A snippet
```
SampleLocaleDenver (cold cache) - 21% slow down
SampleLocaleDenver (warm cache) -15% slow down
SampleLocaleMelbourne (cold cache) - 5% improvement
SampleLocaleMelbourne (warm cache) - about the same
GoogleTiles.LocaleDeathValley (cold cache) - 32% improvement
GoogleTiles.LocaleDeathValley (warm cache) - 15% slow down
GoogleTiles.LocaleChrysler (cold cache) - 25% improvement
GoogleTiles.LocaleChrysler (warm cache) - 9% slow down
```

**Comment by kring** [comment]
I'm seeing frequent crashes just by press Play and flying around in the `01_CesiumWorld` level.

Here's one example:

```
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!CesiumUtility::ReferenceCounted<CesiumRasterOverlays::RasterOverlayTileProvider,0>::addReference() Line 68	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!CesiumUtility::IntrusivePointer<CesiumRasterOverlays::RasterOverlayTileProvider>::addReference() Line 187	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!CesiumUtility::IntrusivePointer<CesiumRasterOverlays::RasterOverlayTileProvider>::IntrusivePointer<CesiumRasterOverlays::RasterOverlayTileProvider>(CesiumRasterOverlays::RasterOverlayTileProvider * p) Line 24	C++
>	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TilesetContentManager::dispatchRasterWork(Cesium3DTilesSelection::RasterProcessingData & processingData, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & responseDataMap, Cesium3DTilesSelection::TileWorkManager::Work * work) Line 1802	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::<lambda>(Cesium3DTilesSelection::RasterProcessingData & processingData, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & responseDataMap, Cesium3DTilesSelection::TileWorkManager::Work * work) Line 913	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::invoke<void <lambda>(Cesium3DTilesSelection::RasterProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) &,Cesium3DTilesSelection::RasterProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>(Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::void <lambda>(Cesium3DTilesSelection::RasterProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) & _Obj, Cesium3DTilesSelection::RasterProcessingData & _Arg1, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args2_0>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Args2_1>) Line 1587	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Invoker_ret<void>::_Call<void <lambda>(Cesium3DTilesSelection::RasterProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) &,Cesium3DTilesSelection::RasterProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>(Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::void <lambda>(Cesium3DTilesSelection::RasterProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) & _Func, Cesium3DTilesSelection::RasterProcessingData & <_Vals_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Vals_1>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Vals_2>) Line 674	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Func_impl_no_alloc<void <lambda>(Cesium3DTilesSelection::RasterProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *),void,Cesium3DTilesSelection::RasterProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>::_Do_call(Cesium3DTilesSelection::RasterProcessingData & <_Args_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args_1>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Args_2>) Line 834	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Func_class<void,Cesium3DTilesSelection::RasterProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>::operator()(Cesium3DTilesSelection::RasterProcessingData & <_Args_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args_1>, Cesium3DTilesSelection::TileWorkManager::Work * <_Args_2>) Line 875	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TileWorkManager::transitionProcessing(std::shared_ptr<Cesium3DTilesSelection::TileWorkManager> & thiz) Line 558	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TileWorkManager::transitionRequests::__l23::<lambda>() Line 476	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!async::detail::invoke_fake_void<void <lambda>(void),void>(Cesium3DTilesSelection::TileWorkManager::transitionRequests::__l23::void <lambda>(void) && f) Line 103	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!async::detail::invoke_fake_void<void <lambda>(void)>(Cesium3DTilesSelection::TileWorkManager::transitionRequests::__l23::void <lambda>(void) && f, async::detail::fake_void __formal) Line 113	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!async::detail::continuation_exec_func<CesiumAsync::CesiumImpl::ImmediateScheduler<CesiumAsync::CesiumImpl::QueuedScheduler>,async::task<void>,async::detail::fake_void,void <lambda>(void),std::integral_constant<bool,1>,0>::operator()(async::detail::task_base * t) Line 550	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!async::detail::task_func<CesiumAsync::CesiumImpl::ImmediateScheduler<CesiumAsync::CesiumImpl::QueuedScheduler>,async::detail::continuation_exec_func<CesiumAsync::CesiumImpl::ImmediateScheduler<CesiumAsync::CesiumImpl::QueuedScheduler>,async::task<void>,async::detail::fake_void,void <lambda>(void),std::integral_constant<bool,1>,0>,async::detail::fake_void>::run(async::detail::task_base * t) Line 398	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!async::task_run_handle::run() Line 134	C++
```

The immediate problem here appears to be that in `TilesetContentManager::dispatchRasterWork`, `pRasterTile->getLoadingTile()` is returning an invalid pointer.

image


Here's another example:

```
>	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Func_class<CesiumAsync::Future<Cesium3DTilesSelection::TileLoadResult>,Cesium3DTilesSelection::TileLoadInput const &,Cesium3DTilesSelection::TilesetContentLoader *>::operator()(const Cesium3DTilesSelection::TileLoadInput & <_Args_0>, Cesium3DTilesSelection::TilesetContentLoader * <_Args_1>) Line 873	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TilesetContentManager::dispatchTileWork(Cesium3DTilesSelection::TileProcessingData & processingData, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & responseDataMap, Cesium3DTilesSelection::TileWorkManager::Work * work) Line 1692	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::<lambda>(Cesium3DTilesSelection::TileProcessingData & processingData, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & responseDataMap, Cesium3DTilesSelection::TileWorkManager::Work * work) Line 905	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::invoke<void <lambda>(Cesium3DTilesSelection::TileProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) &,Cesium3DTilesSelection::TileProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>(Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::void <lambda>(Cesium3DTilesSelection::TileProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) & _Obj, Cesium3DTilesSelection::TileProcessingData & _Arg1, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args2_0>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Args2_1>) Line 1587	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Invoker_ret<void>::_Call<void <lambda>(Cesium3DTilesSelection::TileProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) &,Cesium3DTilesSelection::TileProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>(Cesium3DTilesSelection::TilesetContentManager::createWorkManager::__l2::void <lambda>(Cesium3DTilesSelection::TileProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *) & _Func, Cesium3DTilesSelection::TileProcessingData & <_Vals_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Vals_1>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Vals_2>) Line 674	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Func_impl_no_alloc<void <lambda>(Cesium3DTilesSelection::TileProcessingData &, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> &, Cesium3DTilesSelection::TileWorkManager::Work *),void,Cesium3DTilesSelection::TileProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>::_Do_call(Cesium3DTilesSelection::TileProcessingData & <_Args_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args_1>, Cesium3DTilesSelection::TileWorkManager::Work * && <_Args_2>) Line 834	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!std::_Func_class<void,Cesium3DTilesSelection::TileProcessingData &,std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> const &,Cesium3DTilesSelection::TileWorkManager::Work *>::operator()(Cesium3DTilesSelection::TileProcessingData & <_Args_0>, const std::map<std::string,CesiumAsync::ResponseData,std::less<std::string>,std::allocator<std::pair<std::string const ,CesiumAsync::ResponseData>>> & <_Args_1>, Cesium3DTilesSelection::TileWorkManager::Work * <_Args_2>) Line 875	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TileWorkManager::transitionProcessing(std::shared_ptr<Cesium3DTilesSelection::TileWorkManager> & thiz) Line 554	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TileWorkManager::TryDispatchProcessing(std::shared_ptr<Cesium3DTilesSelection::TileWorkManager> & thiz) Line 249	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::TilesetContentManager::processLoadRequests(std::vector<Cesium3DTilesSelection::TileLoadRequest,std::allocator<Cesium3DTilesSelection::TileLoadRequest>> & requests, Cesium3DTilesSelection::TilesetOptions & options) Line 963	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::Tileset::_processWorkerThreadLoadQueue() Line 1454	C++
 	UnrealEditor-CesiumRuntime-Win64-DebugGame.dll!Cesium3DTilesSelection::Tileset::updateView(const std::vector<Cesium3DTilesSelection::ViewState,std::allocator<Cesium3DTilesSelection::ViewState>> & frustums, float deltaTime) Line 396	C++
```

Here we're in `dispatchTileWork` and the immediate problem is that `loaderCallback` is empty. The assert didn't trigger, though, probably because they're disabled in Unreal.

image

File: Cesium3DTilesSelection/src/TileWorkManager.cpp [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+  std::vector<Work*>& requestWorkVec = foundIt->second;
+  for (Work* requestWork : requestWorkVec) {
+    CesiumAsync::RequestData* workNextRequest = requestWork->getNextRequest();
+    assert(workNextRequest);
+    assert(pCompletedRequest->url() == workNextRequest->url);
+
+    assert(_ownedWork.find(requestWork->uniqueId) != _ownedWork.end());
+
+    // A response code of 0 is not a valid HTTP code
+    // and probably indicates a non-network error.
```
Comment (kring) [inline Cesium3DTilesSelection/src/TileWorkManager.cpp:276]: No, it probably indicates a valid response from a `file:///` URL. It's annoying that libcurl returns a status code of 0 for a successful file read, but it does.

**Comment by csciguy8** [comment]
> I'm seeing frequent crashes just by press Play and flying around in the `01_CesiumWorld` level.

Those crashes look pretty severe and not obvious.

I can't seem to reproduce them though. I checked several of the sample levels in PIE mode and everything seems to run fine. Ran in these 4 configurations...

- Desktop - Windows 10, Unreal 5.4.1, x64-Debug (DebugGame Editor)
- Desktop - Windows 10, Unreal 5.4.1, x64-Release (Development Editor)
- Laptop  - Windows 10, Unreal 5.1.1, x64-Debug (DebugGame Editor)
- Laptop  - Windows 10, Unreal 5.1.1, x64-Release (Development Editor)

(this PR with cesium-unreal main)

Could there be something wrong with your set up?

**Comment by kring** [comment]
> Could there be something wrong with your set up?

I don't have any reason to think so @csciguy8. But I'll try main to make sure it doesn't happen there. They look like race conditions to me, and I was able to reproduce various crashes in both Debug Editor (with Debug build of cesium-native) and Development Editor (with RelWithDebInfo build of cesium-native).

**Comment by kring** [comment]
I switched to main, flew around a whole bunch, and never saw a crash. Switched back to this branch and saw it within seconds. I can easily make it happen at will. I recorded a video, which I'll send to you separately.

**Comment by csciguy8** [comment]
> I switched to main, flew around a whole bunch, and never saw a crash. Switched back to this branch and saw it within seconds. I can easily make it happen at will. I recorded a video, which I'll send to you separately.

@kring Ready for another look regarding those crashes. 

Thanks for the test. Looks like I never tested in cases where tiles would get to the unloading state faster than they would load.

**Comment by electrum-bowie** [comment]
Any progress on this ?

I can't wait for these improvements to come !

**Comment by kring** [comment]
This is on hold for the time being, due to other higher-priority work and because the performance we saw with this approach was a bit mixed (some things got notably faster, but others got slower).

**Comment by electrum-bowie** [comment]
Is there any other work being done for performance ?
